lib/internal/streams/destroy.js

```diff
@@ -252,6 +252,25 @@ function constructNT(stream) {
       process.nextTick(emitConstructNT, stream);
     }
   });
+  if (result !== undefined && result !== null) {
+    try {
+      const then = result.then;
+      if (typeof then === 'function') {
+        then.call(result, undefined, function(err) {
```

Comment:
shouldn't we make the `callback` not required with an async signature? i.e. move the callback into "then". Seems weird to have a mix of both.

Reply:
Yes, I was thinking the same. We can benchmark `_read()` and `_write()` with this, but given that the branch will only be taken if those happen to return a `Promise` I'm not convinced it will be a big performance hit overall.